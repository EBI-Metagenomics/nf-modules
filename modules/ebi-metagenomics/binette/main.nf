process BINETTE {
    tag "$meta.id"
    label 'process_high'

    conda "${moduleDir}/environment.yml"
    container "${workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container
        ? 'https://depot.galaxyproject.org/singularity/binette:1.1.2--pyh7e72e81_0'
        : 'biocontainers/binette:1.1.2--pyh7e72e81_0'}"

    input:
    tuple val(meta), path(input_binning, stageAs: "input_binning/*"), path(contigs), path(proteins)
    val input_type
    path checkm2_db

    output:
    tuple val(meta), path("${meta.id}_final_bins/*")                  , emit: refined_bins
    tuple val(meta), path("${meta.id}_final_bins_quality_reports.tsv"), emit: refined_bins_report
    path "versions.yml"                                               , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args        = task.ext.args ?: ''
    def protein_arg = proteins ? "--proteins ${proteins}" : ""
    def input_arg   = ""

    if (input_type == 'fasta') {
        input_arg = "--bin_dirs"
    } else if (input_type == 'tsv') {
        input_arg = "--contig2bin_tables"
    } else {
        error "Invalid input_type: ${input_type}. Must be 'fasta' or 'tsv'"
    }

    """
    binette \\
        ${input_arg} input_binning/* \\
        ${protein_arg} \\
        --contigs ${contigs} \\
        --threads ${task.cpus} \\
        --checkm2_db ${checkm2_db} \\
        -o . \\
        ${args}

    mv final_bins "${meta.id}_final_bins"
    mv final_bins_quality_reports.tsv ${meta.id}_final_bins_quality_reports.tsv

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        binette: \$( binette --version )
    END_VERSIONS
    """

    stub:
    """
    mkdir ${meta.id}_final_bins
    touch ${meta.id}_final_bins/bin_1.fasta
    touch ${meta.id}_final_bins_quality_reports.tsv

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        binette: \$( binette --version )
    END_VERSIONS
    """
}
