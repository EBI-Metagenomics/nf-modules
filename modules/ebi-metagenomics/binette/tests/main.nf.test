nextflow_process {

    name "Test Process BINETTE"
    tag "modules_ebimetagenomics"
    tag "modules"
    tag "binette"
    script "../main.nf"
    process "BINETTE"


    test("Should refine bins from FASTA input directories") {
        when {
            params {
                outdir = "${launchDir}/tests/results"
            }
            process {
                """
                input[0] = [
                    [id: 'test'],
                    [
                        file('$moduleDir/tests/data/binner1', checkIfExists: true),
                        file('$moduleDir/tests/data/binner2', checkIfExists: true)
                    ],
                    file('$moduleDir/tests/data/metagenome.fasta', checkIfExists: true),
                    []
                ]
                input[1] = "fasta"
                input[2] = file('$moduleDir/tests/data/small_db.dmnd', checkIfExists: true)
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                {
                    assert snapshot(
                        process.out.refined_bins,
                        process.out.refined_bins_report,
                        process.out.versions,
                        path(process.out.versions[0]).yaml
                    ).match()
                }
            )
        }
    }


    test("Should refine bins from TSV contig2bin tables") {
        when {
            params {
                outdir = "${launchDir}/tests/results"
            }
            process {
                """
                input[0] = [
                    [id: 'test'],
                    [
                        file('$moduleDir/tests/data/binner1.tsv', checkIfExists: true),
                        file('$moduleDir/tests/data/binner2.tsv', checkIfExists: true)
                    ],
                    file('$moduleDir/tests/data/metagenome.fasta', checkIfExists: true),
                    []
                ]
                input[1] = "tsv"
                input[2] = file('$moduleDir/tests/data/small_db.dmnd', checkIfExists: true)
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                {
                    assert snapshot(
                        process.out.refined_bins,
                        process.out.refined_bins_report,
                        process.out.versions,
                        path(process.out.versions[0]).yaml
                    ).match()
                }
            )
        }
    }


    test("test_binette - stub") {

        options '-stub'

        when {
            params {
                outdir = "${launchDir}/tests/results"
            }
            process {
                """
                input[0] = [
                    [id: 'test'],
                    [
                        file('$moduleDir/tests/data/binner1', checkIfExists: true),
                        file('$moduleDir/tests/data/binner2', checkIfExists: true)
                    ],
                    file('$moduleDir/tests/data/metagenome.fasta', checkIfExists: true),
                    []
                ]
                input[1] = "fasta"
                input[2] = file('$moduleDir/tests/data/small_db.dmnd', checkIfExists: true)
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                {
                    assert snapshot(
                        process.out.refined_bins,
                        process.out.refined_bins_report,
                        process.out.versions,
                        path(process.out.versions[0]).yaml
                    ).match()
                }
            )
        }
    }
}