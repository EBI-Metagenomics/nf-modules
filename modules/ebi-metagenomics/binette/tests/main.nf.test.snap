{
    "test_binette - stub": {
        "content": [
            [
                [
                    {
                        "id": "test"
                    },
                    "bin_1.fasta:md5,d41d8cd98f00b204e9800998ecf8427e"
                ]
            ],
            [
                [
                    {
                        "id": "test"
                    },
                    "test_final_bins_quality_reports.tsv:md5,d41d8cd98f00b204e9800998ecf8427e"
                ]
            ],
            [
                "versions.yml:md5,a733c14a8248c5bd038523fd1a5ebc10"
            ],
            {
                "BINETTE": {
                    "binette": "1.1.2"
                }
            }
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "25.04.1"
        },
        "timestamp": "2025-07-28T16:26:24.512245"
    },
    "Should refine bins from FASTA input directories": {
        "content": [
            [
                [
                    {
                        "id": "test"
                    },
                    "bin_5.fa:md5,64cc2172b00c5accd60f10bbc96ea91a"
                ]
            ],
            [
                [
                    {
                        "id": "test"
                    },
                    "test_final_bins_quality_reports.tsv:md5,8b2381fadab2172048e56b86cabdf13a"
                ]
            ],
            [
                "versions.yml:md5,a733c14a8248c5bd038523fd1a5ebc10"
            ],
            {
                "BINETTE": {
                    "binette": "1.1.2"
                }
            }
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "25.04.1"
        },
        "timestamp": "2025-07-28T16:27:28.552564"
    },
    "Should refine bins from TSV contig2bin tables": {
        "content": [
            [
                [
                    {
                        "id": "test"
                    },
                    "bin_5.fa:md5,64cc2172b00c5accd60f10bbc96ea91a"
                ]
            ],
            [
                [
                    {
                        "id": "test"
                    },
                    "test_final_bins_quality_reports.tsv:md5,8b2381fadab2172048e56b86cabdf13a"
                ]
            ],
            [
                "versions.yml:md5,a733c14a8248c5bd038523fd1a5ebc10"
            ],
            {
                "BINETTE": {
                    "binette": "1.1.2"
                }
            }
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "25.04.1"
        },
        "timestamp": "2025-07-28T16:26:14.508337"
    }
}